### DISCUSSION TITLE
A rather large batch - guess finish time...

### DISCUSSION THREAD

**Discussion by stefankinell** [karttapullautin/karttapullautin#170] [Show and tell]
Today I am starting a rather large batch.

12 707 laz files of 2,5 x 2,5 km. Plus 991 bordering files. This totals 79,418.75 square kilometers - about 18 % of Swedens total area.

Batch started 08:09:29 today Saturday, one computer - Macbook M1 Max. What is your guesses of when it will be finished?

![image](https://github.com/user-attachments/assets/77bfaf5c-da38-494d-804c-adf37926bfa6)

**Comment by rphlo** [comment]
I'd say sometime Tuesday, I assume there is no Shapefile rendering...

**Comment by rphlo** [comment]
I guess you didnt pick the experimental RAM filesystem...

**Comment by stefankinell** [comment]
I think i took one of the latest, if not the latest releases. I assumed that would be working :-). 

No shape files. Standard processing of contours, vegetation and cliffs. 

Let's see how it endures. It's flint fastet than i expected to start with. But it is the high mountains of Sweden so the most cliff intensive area to process :-D

**Comment by antbern** [comment]
I'll go for a more optimistic estimation: Sunday some time between 20:00 and 22:00 🤞 🚀

**Comment by Jagge** [comment]
I have been processing for days a 50 000 km2 area with my laptop and it apperently takes 10..12 days. But my bottleneck is my internet connection bandwidth, KP with just half of the cores assigned still prosesses faster than I download tiles.  The prosess kind of streams tiles by downloading, prosecessing and then deleting laz files on the fly. So I'd say that Sweden batch will take maybe 6 days? I have been using slightly tweaked 2.4.7. Tiles are 1x1 km the prosess has been spitting out 3 tiles/min on average for some days now (= same as my tile download speed).

**Reply by stefankinell** [reply]
I had a similar issue, and did all the download first. It took days and disconnects from Lantmäteriet - also talks with them if their ftp was having issues. Finally I realized it was my TP-link DECO router that classified FTP as a dangerous protocol and stopped it mid-download. When I noticed that and found a workaround I could download all within a few hours.

**Reply by Jagge** [reply]
I am downloading over https. My home internet just is a bit slow and I don't want to make it unusable for other purposes. I check with lasinfo if the laz file is ok and try again later if it is not, sometimes (my laz file is a text file with an error message). I don't have enough free disk space to download everything (several terabytes), so I download row by row so that I always have 3 rows for precessing, process just the middle one while downloading new fourth one and after processin delete the row not needed. Like that I can just keep it rolling and process as several one row batches (neighboring tile rows faked as already processed tiles).

Tweaks are several fixes for 2.5m interval support (needed because I processing relatively flat area), own median filter settings for yellow and some minor bug fixes.

**Reply by stefankinell** [reply]
Smart with the dynamic downloading so you get bordering tiles available when processing. I have plenty of HD space in servers locally, so I instead have a process of fetching bordering tiles when I now do processing, and then do square by square. Also preprocessing with downloading internal to working machines before it is time to process.

are you using a different method to do 2,5 meter intervalls than the one we introduced earlier this year with variable contour height? I am about to do 2,5 meters contours as well for the flatter parts of Sweden as a complimentary layer. Would be interesting to hear if you have another solution that is better than the one we have in the tool now. Also for the yellow filtering - this is on my todo want list but have not sorted out other more prioritized things first :-D Like this weeks task to set up a geoserver to dynamically be able to generate some map layers for coming projects. But that is not really KP-related...

**Reply by Jagge** [reply]
method I use for 2.5m is the one there already it, but with bug fixes and some improvements. The current version still has several hardcoded values for 5m baked in. I also added some additional logics to not draw them when its too steep. It's still nowhere near perfect but better. 
![470579840_636486452036563_3608350092511487956_n](https://github.com/user-attachments/assets/e40a60e9-5971-4663-99e3-6fc216f51331)

**Reply by stefankinell** [reply]
"additional logics to not draw them when its too steep" that sounds like the (in)famous Stockholm contours that on the map was defined as "between 1 and 4 meters contour intervalls".

But the yellow looks smooth and nice on this picture. I tried to use the smoothed yellow rendering that is available in current releases but it smoothed to much for my opinion. One day when it is time to redo vegetation for all Sweden again I want to look into how I can do that better, like differentiated smoothing for green and yellow - or just a possibility to have the same yellow areas but rounder than the squared ones you get from the standard settings, if not for anything else - for a better visual appearance.

**Reply by rphlo** [reply]
We are impatiently waiting for the PR @Jagge !!!

**Reply by Jagge** [reply]
I know, I'll do it as soon as the current batch gets done and my laptop becomes not too hot to touch again.

**Comment by stefankinell** [comment]
And now I am done (or actually yesterday was a more fair statement of when It was done) I had to restart the script a couple of times in the end when it crashed. But a restart of the computer and deletion of empty .png files in the out-folder made it work again.

So some data. 

Started on the 14th of December 8:20
Finished on the 17th of December 10:30 (calculated when I deducted the stops)
Total time: 74 hours, 10 minutes.

The progression of number of ready files
![image](https://github.com/user-attachments/assets/c3e996af-a9b7-4235-955b-914c3540e7a1)

The spread of processing time per file. (recalculated per CPU core.)
![image](https://github.com/user-attachments/assets/0740f4a1-c0c4-49b5-84ca-56806e527867)

Green bars are frequency in brackets of 15 seconds. Blue line is cumulated numbers of the frequency count. Which shows that 25% of the files were processed under a minute each.
 
Average processing time per square: 3 minutes 34 seconds
Median processing time per square: 2 minutes 40 seconds.

There were some outliers that took plenty of time to process. so that is why the average time is higher than the median.

Regarding processing time - this was a batch to process the most hilly area of Sweden. So this is most definitely higher values than what would be the average of the total country.

**Reply by rphlo** [reply]
What do I win with my correct prediction?

Can you tell what was the reason for the crashes? Bad LAZ files?

**Reply by stefankinell** [reply]
You have earned the titel "Grand Pronostiqueur des Délais" that you may sign any forthcoming posts with.

I can not find my way back to the output of the crashes. I saved them but then when it worked after the restart of the computer I must have closed that document without saving. So the laz-files probably were OK.

**Comment by rphlo** [comment]
Do you have a timeline when the tiles will be available on kartor.gokartor.se?

**Reply by stefankinell** [reply]
Right now I do not not know if or when. These lidar-files are not open data yet. I can work with them since I have a partnership deal with Lantmäteriet. Which mean I can work with the data and produce products - which e.g. governmental agencies can use without cost for the data - but private persons and companies have to pay licenses for. For my company it is way to costly to purchase all data to publish on the map site. But it helps me create products that governmental agencies can use - and I can also sell extracts of the map for not so high costs to e.g. orienteers that like to have a orienteering map when hiking in the Swedish mountains.

But there is hope that this data will be open in the future so that it this part can be published open.
